server/src/test/java/org/apache/druid/server/coordinator/simulate/CoordinatorSimulationBaseTest.java

```diff
@@ -0,0 +1,299 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.server.coordinator.simulate;
+
+import org.apache.druid.client.DruidServer;
+import org.apache.druid.java.util.common.granularity.Granularities;
+import org.apache.druid.java.util.emitter.core.Event;
+import org.apache.druid.server.coordination.ServerType;
+import org.apache.druid.server.coordinator.CoordinatorDynamicConfig;
+import org.apache.druid.server.coordinator.CreateDataSegments;
+import org.apache.druid.server.coordinator.rules.ForeverLoadRule;
+import org.apache.druid.server.coordinator.rules.Rule;
+import org.apache.druid.timeline.DataSegment;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Base test for coordinator simulations.
+ * <p>
+ * Each test must call {@link #startSimulation(CoordinatorSimulation)} to start
+ * the simulation. {@link CoordinatorSimulation#stop()} should not be called as
+ * the simulation is stopped when cleaning up after the test in {@link #tearDown()}.
+ * <p>
+ * Tests that verify balancing behaviour should set
+ * {@link CoordinatorDynamicConfig#useBatchedSegmentSampler()} to true.
+ * Otherwise, the segment sampling is random and can produce repeated values
+ * leading to flakiness in the tests. The simulation sets this field to true by
+ * default.
+ */
+public abstract class CoordinatorSimulationBaseTest
```

Comment:
Consider making this a "fixture" rather than a base test. With a fixture, you write a test like so:

```java
class MyTest
{
  CoordinatorSimulationFixture fixture = new CoordinatorSimulationFixture(...);

  @Before
  public void setup()
  {
    fixture.setup();
    // My own setup
  }

  @Test
  public void myTest()
  {
    fixture.startSimulation(...);
    fixture.doOtherStuff(...);
  }
```

The point is that the test class is simple. The test writer can pull in other dependencies without having a multiple-inheritance problem. Etc.

Reply:
You could also have the fixture and then also have a BaseTest implemented using the fixture.  Then you are effectively actually using a composable fixture for things (and people can fall back to that if need be), but still don't have to repeat `fixture.` in all of the places.

Reply:
Agreed. Although, I think the `CoordinatorSimulation` object itself is already (somewhat) fulfilling the role of the fixture.

It's easy for a developer to write simulation-based tests without ever having to extend the `BaseTest`. The `BaseTest` just provides a bunch of convenience methods which do nothing but delegate to the simulation itself.

In the current tests, other than action invocations on the simulation object, 
each test does only the following:
1. declare inputs to the simulation, which would be different for each test case
2. extract and map metrics from the simulation, which can be commoned out into the simulation itself
3. verifications

I will make the updates for 2 thus making the `BaseTest` an even thinner layer on top of the simulation.
But, as there is hardly any common setup required for the tests, I am not sure if we need a fixture just yet.